### PR TITLE
=str #23933 materialization failing hard with nicer message now

### PR DESCRIPTION
**Pull request by ktoso**
Refs https://github.com/akka/akka/issues/23933

Please read the comment https://github.com/akka/akka/issues/23933#issuecomment-343897487 to get the rationale behind not going in deeper into this minefield

### CODE REVIEW

**Comment by ktoso** [comment]
The improved message can be beneficial on `master` as well; so if you think this is good, I'll submit it to master as well

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala [hunk 1]
```diff
@@ -812,24 +812,35 @@ private[impl] class VirtualPublisher[T] extends AtomicReference[AnyRef] with Pub
...
+        if (!compareAndSet(null, pub)) registerPublisher(pub) // retry
+
       case sub: Subscriber[r] ⇒
         set(Inert.subscriber)
         pub.asInstanceOf[Publisher[r]].subscribe(sub)
-      case _ ⇒ throw new IllegalStateException("internal error")
+
+      case p: Publisher[_] ⇒
+        throw new IllegalStateException(s"internal error, already registered [$p], yet attempted to register 2nd publisher [$pub]!")
+
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: Looks like scalariform doesn't like the whitespace here..

**Review by raboof — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala (hunk 1 above)
Comment (ktoso) [reply akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala]: :( Will fix. Did not reload sbt when switching to old branch I think so `master`

**Review by patriknw — APPROVED** [review]
lgtm

**Comment by ktoso** [comment]
Note to self to port it to master roo

**Comment by ktoso** [comment]
PLS BUILD

**Pull request merged**